[DataLoader] Typing Enforcement for DataPipe at runtime

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54773 [DataLoader] Typing Doc
* **#54544 [DataLoader] Typing Enforcement for DataPipe at runtime**
* #54066 [DataLoader] Typing Enforcement for DataPipe at construct-time
* #54020 [DataLoader] Typing Enforcement for DataPipe at Compile-time
* #54299 [DataLoader] Implement issubtype for type hints

## Feature
- Add `subinstance(data, type)` to check `data` is a subtype instance of the `type`
- Add a decorator of `runtime_validation` to validate the returned data from `__iter__` is subtype instance of hint.

## Next step
- Add context manager for runtime checking
- Pass runtime checking after one epoch
- Interaction with `compile-time` static checking.

Differential Revision: [D27327234](https://our.internmc.facebook.com/intern/diff/D27327234)

ejguan committed:
[DataLoader] Typing Enforcement for DataPipe at runtime

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at runtime"


- Add `subinstance(data, type)`

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at runtime"


- Add `subinstance(data, type)`

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at runtime"


- Add `subinstance(data, type)`

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at runtime"


## Feature
- Add `subinstance(data, type)` to check `data` is a subtype instance of the `type`
- Add a decorator of `runtime_validation` to validate the returned data from `__iter__` is subtype instance of hint.

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at runtime"


## Feature
- Add `subinstance(data, type)` to check `data` is a subtype instance of the `type`
- Add a decorator of `runtime_validation` to validate the returned data from `__iter__` is subtype instance of hint.

## Next step
- Add context manager for runtime checking
- Pass runtime checking after one epoch
- Interaction with `compile-time` static checking.

Differential Revision: [D27327234](https://our.internmc.facebook.com/intern/diff/D27327234)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at runtime"


## Feature
- Add `subinstance(data, type)` to check `data` is a subtype instance of the `type`
- Add a decorator of `runtime_validation` to validate the returned data from `__iter__` is subtype instance of hint.

## Next step
- Add context manager for runtime checking
- Pass runtime checking after one epoch
- Interaction with `compile-time` static checking.

Differential Revision: [D27327234](https://our.internmc.facebook.com/intern/diff/D27327234)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at runtime"


## Feature
- Add `subinstance(data, type)` to check `data` is a subtype instance of the `type`
- Add a decorator of `runtime_validation` to validate the returned data from `__iter__` is subtype instance of hint.

## Next step
- Add context manager for runtime checking
- Pass runtime checking after one epoch
- Interaction with `compile-time` static checking.

Differential Revision: [D27327234](https://our.internmc.facebook.com/intern/diff/D27327234)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at runtime"


## Feature
- Add `subinstance(data, type)` to check `data` is a subtype instance of the `type`
- Add a decorator of `runtime_validation` to validate the returned data from `__iter__` is subtype instance of hint.

## Next step
- Add context manager for runtime checking
- Pass runtime checking after one epoch
- Interaction with `compile-time` static checking.

Differential Revision: [D27327234](https://our.internmc.facebook.com/intern/diff/D27327234)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at runtime"


## Feature
- Add `subinstance(data, type)` to check `data` is a subtype instance of the `type`
- Add a decorator of `runtime_validation` to validate the returned data from `__iter__` is subtype instance of hint.

## Next step
- Add context manager for runtime checking
- Pass runtime checking after one epoch
- Interaction with `compile-time` static checking.

Differential Revision: [D27327234](https://our.internmc.facebook.com/intern/diff/D27327234)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at runtime"


## Feature
- Add `subinstance(data, type)` to check `data` is a subtype instance of the `type`
- Add a decorator of `runtime_validation` to validate the returned data from `__iter__` is subtype instance of hint.

## Next step
- Add context manager for runtime checking
- Pass runtime checking after one epoch
- Interaction with `compile-time` static checking.

Differential Revision: [D27327234](https://our.internmc.facebook.com/intern/diff/D27327234)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at runtime"


## Feature
- Add `subinstance(data, type)` to check `data` is a subtype instance of the `type`
- Add a decorator of `runtime_validation` to validate the returned data from `__iter__` is subtype instance of hint.

## Next step
- Add context manager for runtime checking
- Pass runtime checking after one epoch
- Interaction with `compile-time` static checking.

Differential Revision: [D27327234](https://our.internmc.facebook.com/intern/diff/D27327234)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at runtime"


## Feature
- Add `subinstance(data, type)` to check `data` is a subtype instance of the `type`
- Add a decorator of `runtime_validation` to validate the returned data from `__iter__` is subtype instance of hint.

## Next step
- Add context manager for runtime checking
- Pass runtime checking after one epoch
- Interaction with `compile-time` static checking.

Differential Revision: [D27327234](https://our.internmc.facebook.com/intern/diff/D27327234)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at runtime"


## Feature
- Add `subinstance(data, type)` to check `data` is a subtype instance of the `type`
- Add a decorator of `runtime_validation` to validate the returned data from `__iter__` is subtype instance of hint.

## Next step
- Add context manager for runtime checking
- Pass runtime checking after one epoch
- Interaction with `compile-time` static checking.

Differential Revision: [D27327234](https://our.internmc.facebook.com/intern/diff/D27327234)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at runtime"


## Feature
- Add `subinstance(data, type)` to check `data` is a subtype instance of the `type`
- Add a decorator of `runtime_validation` to validate the returned data from `__iter__` is subtype instance of hint.

## Next step
- Add context manager for runtime checking
- Pass runtime checking after one epoch
- Interaction with `compile-time` static checking.

Differential Revision: [D27327234](https://our.internmc.facebook.com/intern/diff/D27327234)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at runtime"


## Feature
- Add `subinstance(data, type)` to check `data` is a subtype instance of the `type`
- Add a decorator of `runtime_validation` to validate the returned data from `__iter__` is subtype instance of hint.

## Next step
- Add context manager for runtime checking
- Pass runtime checking after one epoch
- Interaction with `compile-time` static checking.

Differential Revision: [D27327234](https://our.internmc.facebook.com/intern/diff/D27327234)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at runtime"


## Feature
- Add `subinstance(data, type)` to check `data` is a subtype instance of the `type`
- Add a decorator of `runtime_validation` to validate the returned data from `__iter__` is subtype instance of hint.

## Next step
- Add context manager for runtime checking
- Pass runtime checking after one epoch
- Interaction with `compile-time` static checking.

Differential Revision: [D27327234](https://our.internmc.facebook.com/intern/diff/D27327234)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at runtime"


## Feature
- Add `subinstance(data, type)` to check `data` is a subtype instance of the `type`
- Add a decorator of `runtime_validation` to validate the returned data from `__iter__` is subtype instance of hint.

## Next step
- Add context manager for runtime checking
- Pass runtime checking after one epoch
- Interaction with `compile-time` static checking.

Differential Revision: [D27327234](https://our.internmc.facebook.com/intern/diff/D27327234)

[ghstack-poisoned]

VitalyFedyunin reviewed (approved):
Please follow-up with PR to context enable/disable runtime checks.

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at runtime"


## Feature
- Add `subinstance(data, type)` to check `data` is a subtype instance of the `type`
- Add a decorator of `runtime_validation` to validate the returned data from `__iter__` is subtype instance of hint.

## Next step
- Add context manager for runtime checking
- Pass runtime checking after one epoch
- Interaction with `compile-time` static checking.

Differential Revision: [D27327234](https://our.internmc.facebook.com/intern/diff/D27327234)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at runtime"


## Feature
- Add `subinstance(data, type)` to check `data` is a subtype instance of the `type`
- Add a decorator of `runtime_validation` to validate the returned data from `__iter__` is subtype instance of hint.

## Next step
- Add context manager for runtime checking
- Pass runtime checking after one epoch
- Interaction with `compile-time` static checking.

Differential Revision: [D27327234](https://our.internmc.facebook.com/intern/diff/D27327234)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at runtime"


## Feature
- Add `subinstance(data, type)` to check `data` is a subtype instance of the `type`
- Add a decorator of `runtime_validation` to validate the returned data from `__iter__` is subtype instance of hint.

## Next step
- Add context manager for runtime checking
- Pass runtime checking after one epoch
- Interaction with `compile-time` static checking.

Differential Revision: [D27327234](https://our.internmc.facebook.com/intern/diff/D27327234)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at runtime"


## Feature
- Add `subinstance(data, type)` to check `data` is a subtype instance of the `type`
- Add a decorator of `runtime_validation` to validate the returned data from `__iter__` is subtype instance of hint.

## Next step
- Add context manager for runtime checking
- Pass runtime checking after one epoch
- Interaction with `compile-time` static checking.

Differential Revision: [D27327234](https://our.internmc.facebook.com/intern/diff/D27327234)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at runtime"


## Feature
- Add `subinstance(data, type)` to check `data` is a subtype instance of the `type`
- Add a decorator of `runtime_validation` to validate the returned data from `__iter__` is subtype instance of hint.

## Next step
- Add context manager for runtime checking
- Pass runtime checking after one epoch
- Interaction with `compile-time` static checking.

Differential Revision: [D27327234](https://our.internmc.facebook.com/intern/diff/D27327234)

[ghstack-poisoned]
